smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpBindingProtocolGenerator.java

```diff
@@ -284,70 +272,133 @@ private void generateOperationSerializer(
+                    String headerValue = getInputValue(context, binding.getLocation(),
+                            "input." + memberName + "[suffix]", output, binding.getMember(), target);
+                    // Append the suffix to the defined prefix and serialize the value in to that key.
+                    writer.write("headers['$L' + suffix] = $L;", binding.getLocationName(), headerValue);
+                });
+            });
+        }
+    }
+
+    private List<HttpBinding> writeRequestBody(
+            GenerationContext context,
+            OperationShape operation,
+            HttpBindingIndex bindingIndex
+    ) {
+        TypeScriptWriter writer = context.getWriter();
+        List<HttpBinding> documentBindings = bindingIndex.getRequestBindings(operation, Location.DOCUMENT);
+        documentBindings.sort(Comparator.comparing(HttpBinding::getMemberName));
+        List<HttpBinding> payloadBindings = bindingIndex.getRequestBindings(operation, Location.PAYLOAD);
+
+        if (!documentBindings.isEmpty()) {
+            // Write the default `body` property.
+            context.getWriter().write("let body: any = undefined;");
+            serializeDocument(context, operation, documentBindings);
+            return documentBindings;
+        }
+        if (!payloadBindings.isEmpty()) {
+            SymbolProvider symbolProvider = context.getSymbolProvider();
+            // There can only be one payload binding.
+            HttpBinding binding = payloadBindings.get(0);
+            String memberName = symbolProvider.toMemberName(binding.getMember());
+            Shape target = context.getModel().getShapeIndex().getShape(binding.getMember().getTarget()).get();
+            writer.write("let body: any = $L;", getInputValue(
+                    context, Location.PAYLOAD, "input." + memberName, operation, binding.getMember(), target));
+            return payloadBindings;
+        }
+
+        return ListUtils.of();
+    }
+
     protected String getInputValue(
             GenerationContext context,
             Location bindingType,
-            Shape shape,
+            String dataSource,
+            Shape container,
             MemberShape member,
             Shape target
     ) {
-        String memberName = member.getMemberName();
         SymbolProvider symbolProvider = context.getSymbolProvider();
+        if (container instanceof UnionShape && !unionTargetNeedsSerde(context.getModel().getShapeIndex(), target)) {
+            return dataSource;
+        }
 
-        if (target instanceof StringShape || target instanceof DocumentShape) {
-            return "input." + memberName;
-        } else if (target instanceof BooleanShape || target instanceof NumberShape) {
-            // Just toString on the value.
-            return "input." + memberName + ".toString()";
+        if (target instanceof StringShape || target instanceof DocumentShape
```

Comment:
Might be a good idea to break this out into a method that returns a boolean and has a very descriptive name